### PR TITLE
Budget App PR

### PR DESCRIPTION
**Pull request by AmaduKamara**
# Ruby on Rails capstone project - Budget app
In this capstone project, I worked on a budget app that users can create categories and expenses for a category to keep track of their expenses

🧑‍💻🧑‍💻🧑‍💻🧑‍💻🧑‍💻 
## The App Entails the following: 
- Splash screen
- Sign up and login pages
- Home page (categories page)
- Transactions page
- "Add a new category" page
- "Add a new transaction" page

### CODE REVIEW

File: app/controllers/categories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    @total_amount = []
+
+    @categories.each do |category|
+      total = 0
+      category.category_expenses.each do |category_expense|
+        total += category_expense.expense.amount
+      end
+      @total_amount.push(total)
+    end
+  end
```
Comment (MrBrN197) [inline app/controllers/categories_controller.rb:17]: - [x] The functionality for this action is correct although, I've noticed that this `/categories/1/expenses` page is visible even when there is no logged-in user. this page should require a user to be logged in and only that user should be able to see their expenses.

File: app/views/expenses/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    <div class="mt-2">
+    <%= form.label :amount %><br>
+    <%= form.number_field :amount, class: 'mt-2 w-56 p-1 rounded', require: true,  class: 'border py-2 w-full mb-3 px-3', placeholder: 'Enter first name' %>
+    </div>
+
+    <%= link_to 'Back', category_expenses_path(params[:category_id]), class: "border border-sky-500 text-sky-500 rounded mt-4 mr-20 px-8 py-3" %>
+    <%= form.submit "Submit", class: "bg-sky-500 py-3 px-8 rounded text-slate-50 mt-4" %>
+  <% end %>
+</div>
```
Comment (MrBrN197) [inline app/views/expenses/new.html.erb]: - [x] Here you have the name and the Ammount fields as required. however, you also need a field for the category and one category should be required.
- [x] [OPTIONAL] currently, your input fields all have the same placeholder text. please modify these placeholders so they describe their input fields correctly.

**Review by MrBrN197 — CHANGES_REQUESTED** [review]
## Hi,  @AmaduKamara  👋  

You've done a really nice job so far.
There are some issues that you still need to work on to go to the next project but you are almost there!
## To highlight 👏👏
- ✔ great work with the design of the project.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/categories_controller.rb (hunk 1 above)
Comment (AmaduKamara) [reply app/controllers/categories_controller.rb:17]: Ooh yeah. Thanks for highlighting that. 
I have fixed it and now only auth user can access that route and all other routes related to the authenticated user. 👍

File: app/views/expenses/new.html.erb (hunk 2 above)
Comment (AmaduKamara) [reply app/views/expenses/new.html.erb]: In the first comment. A user can create a category or if a category exists, then that user can click it and make a transaction for that category. That's working perfectly and any transaction created for a category is stored for that category only. I believe that does it, my brother. 👍 Thanks for the review.

File: app/views/categories/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  <%= render 'shared/notice' %>
+</div>
+
+<div class="mx-5 items-center mb-10 pt-20">
+  <% @categories.each_with_index do |category, index| %>
+    <%= link_to category_expenses_path(category.id), class: do %>
+    <div class="flex justify-between bg-white rounded-md shadow-lg p-4 mb-6">
+      <div>
+        <%= image_tag category.icon, alt: 'icon', class: "w-24" %>
+      </div>
```
Comment (medaminedev66) [inline app/views/categories/index.html.erb:15]: - [ ] When I try to sign up and I click on the `sign up button, it gives me this error below. I recommend fixing the syntax error so the categories index page opens without errors:

image

More details:

image

**Review by medaminedev66 — CHANGES_REQUESTED** [review]
Hi @AmaduKamara 😃,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- Your code organization was really great

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/expenses/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    <div class="mt-2">
+    <%= form.label :amount %><br>
+    <%= form.number_field :amount, class: 'mt-2 w-56 p-1 rounded', require: true,  class: 'border py-2 w-full mb-3 px-3', placeholder: 'Enter amount' %>
+    </div>
+
+    <%= link_to 'Back', category_expenses_path(params[:category_id]), class: "border border-sky-500 text-sky-500 rounded mt-4 mr-20 px-8 py-3" %>
+    <%= form.submit "Submit", class: "bg-sky-500 py-3 px-8 rounded text-slate-50 mt-4" %>
+  <% end %>
+</div>
```
Comment (medaminedev66) [inline app/views/expenses/new.html.erb]: - [x] As requested by the previous reviewer, you should add a field of categories (mandatory at least once) to the form so the user is able to choose the category or categories for the transaction.

File: app/views/expenses/new.html.erb (hunk 2 above)
Comment (medaminedev66) [reply app/views/expenses/new.html.erb]: Hi @AmaduKamara, I understand your point of view. But following the project requirements, you should create a form with three fields (Name, Amount, and Categories). The transaction will belong to all chosen categories. Check out the requirements below:

image

File: app/views/categories/index.html.erb (hunk 3 above)
Comment (AmaduKamara) [reply app/views/categories/index.html.erb:15]: The Signup Button works well on my end here. I just signed up and created this category in the screenshot below and there was no error. Kindly configure your database. From my database.yml file. I suggest you comment lines 23 and 24 before running `rails db:create`
```mdx
  username: Amkam
  password: amkampgadmin
```

image

File: app/views/expenses/new.html.erb (hunk 4 above)
Comment (AmaduKamara) [reply app/views/expenses/new.html.erb]: Thank you. I've figured it out and I've added the categories field so users can select a category while creating a transaction. 👍

**Review by Cathella — APPROVED** [review]
## 🔥 Status: Approved 🟢

### Hi @AmaduKamara 👋🏻, 

Great job making the requested changes 💪🏻! Your app looks great 😍.
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

rosycheeks-mochi-peach


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**